File: Data Structures and Algorithms/C++ Version/SelectionSort.cpp
#include <iostream>
#include <vector>

// Global new_array to store the sorted elements
std::vector<int> new_array;

void selectionSort(std::vector<int>& l) {
    while (!l.empty()) {
        int element = l.at(0);
        int idx = 0;

        // Find the minimum element and its index
        for(int i = 0; i < l.size(); i++) {
            if (element > l.at(i)) {
                element = l.at(i);
                idx = i;
            }
        }

        // Add the minimum element to new_array
        new_array.push_back(element);

        // Remove the minimum element from l
        l.erase(l.begin() + idx);
    }

    std::cout << "Process finished" << std::endl;
}

int main() {
    std::vector<int> sample_list = {5, 3, 2, 7, 9, 8};
    std::vector<int> copy_list(sample_list.begin(), sample_list.end());

    selectionSort(copy_list);

    std::cout << "Original array: ";
    for (const auto& e : sample_list) {
        std::cout << e << " ";
    }
    std::cout << std::endl;

    std::cout << "Sorted array: ";
    for (const auto& e : new_array) {
        std::cout << e << " ";
    }
    std::cout << std::endl;

    return 0;
}

File: Data Structures and Algorithms/Python version/SelectionSort.py
# Initialize an empty list to store sorted elements
new_arr = []


def selectionSort(l: list):
    """
    Recursively sorts a list using the selection sort algorithm and stores the sorted elements in new_arr.

    Args:
        l (list): The list to be sorted.

    The function works by finding the minimum element in the list, removing it from the list,
    and appending it to new_arr. This process is repeated recursively until the original list is empty.
    """
    # Assume the first element is the minimum
    element = l[0]
    idx = 0

    # Iterate over the list to find the actual minimum element
    for i in range(len(l)):
        if element > l[i]:
            element = l[i]
            idx = i

    # Remove the minimum element from the list and append it to new_arr
    new_arr.append(l.pop(idx))

    # If the list is empty, the sorting process is complete
    if l == []:
        print("Process finished.")
    else:
        # Recursively call selectionSort to sort the remaining elements
        selectionSort(l)


# Sample list to sort
sample_list = [5, 3, 2, 10, 9, 8]

# Call the selectionSort function with a copy of the sample list
selectionSort(sample_list.copy())

# Print the original and sorted lists
print(f"Original list: {sample_list}")
print(f"After Sorting: {new_arr}")